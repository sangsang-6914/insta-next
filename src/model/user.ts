export interface AuthUser {
  username: string;
  name: string;
  image?: string | undefined;
  email: string;
  id: string;
}

export type SimpleUser = Pick<AuthUser, 'username' | 'image'>;

export interface SearchUser extends AuthUser {
  following: number;
  followers: number;
}

export interface ProfileUser extends SearchUser {
  posts: number;
}

export interface HomeUser extends AuthUser {
  following: SimpleUser[];
  followers: SimpleUser[];
  bookmarks: string[];
}
