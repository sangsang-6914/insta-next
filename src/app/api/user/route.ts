import { getSearchUser } from '@/sanity/user';
import { NextRequest, NextResponse } from 'next/server';

export async function GET(_: NextRequest) {
  return getSearchUser()
    .then((data) => NextResponse.json(data))
    .catch((err) => new Response(`Server Error : ${err}`, { status: 400 }));
}
