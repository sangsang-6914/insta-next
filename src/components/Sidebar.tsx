'use client';

import { AuthUser } from '@/model/user';
import React from 'react';
import Avatar from './Avatar';

interface Props {
  user: AuthUser;
}

function Sidebar({ user }: Props) {
  return (
    <section className="flex flex-col gap-4">
      <div className="flex items-center">
        <Avatar size="medium" image={user.image || ''} />
        <div className="flex flex-col ml-5">
          <p className="font-bold">{user.username}</p>
          <p>{user.name}</p>
        </div>
      </div>
      <div>
        <p className="text-neutral-500">
          About ∙ Help ∙ Press ∙ API ∙ Jobs ∙ <br />
          Privacy ∙ Terms ∙ Location ∙ <br />
          Language
        </p>
      </div>
      <div>
        <p className="text-neutral-500 font-bold">
          @Copyright INSTANTGRAM <br /> from METAL
        </p>
      </div>
    </section>
  );
}

export default Sidebar;
