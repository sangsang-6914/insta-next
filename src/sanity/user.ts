import { ProfileUser, SearchUser } from '@/model/user';
import { client } from './sanity';

export async function getSearchUser(keyword?: string) {
  const condition = keyword
    ? `&& username match "${keyword}" || name match "${keyword}"`
    : '';
  const qr = `
    *[_type == 'user' ${condition}][]{
      ...,
      "id": _id,
      "followers": count(followers),
      "following": count(following),
    }
  `;
  return client.fetch(qr).then((users) =>
    users.map((user: SearchUser) => ({
      ...user,
      followers: user.followers ?? 0,
      following: user.following ?? 0,
    }))
  );
}

export async function getUserProfile(username: string) {
  const qr = `
    *[_type == 'user' && username == "${username}"][0]{
      ...,
      "followers": count(followers),
      "following": count(following),
      "posts": count(*[_type=='post' && author->username == "${username}"])
    }
  `;

  return client.fetch(qr).then((user: ProfileUser) => ({
    ...user,
    followers: user.followers ?? 0,
    following: user.following ?? 0,
    posts: user.posts ?? 0,
  }));
}
