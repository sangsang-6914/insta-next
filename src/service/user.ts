import { client } from '@/sanity/sanity';

export async function getUserByUsername(username: string) {
  const query = `
    *[_type=='user' && username == "${username}"][0]{..., "id": _id, username, name, email, following[]->{username,image}}
  `;

  return client.fetch(query);
}
