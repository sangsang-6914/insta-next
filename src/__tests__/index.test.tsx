import Sidebar from '@/components/Sidebar';
import { AuthUser } from '@/model/user';
import { render } from '@testing-library/react';

describe('Component Test', () => {
  const user: AuthUser = {
    username: 'test',
    email: 'test@gmail.com',
    name: 'test',
    id: '1234',
    image: undefined,
  };
  it('Sidebar render test', () => {
    const { container } = render(<Sidebar user={user} />);
    expect(container).toMatchSnapshot();
  });
});
