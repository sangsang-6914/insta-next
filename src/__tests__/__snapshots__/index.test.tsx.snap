// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Test Sidebar render test 1`] = `
<div>
  <section
    class="flex flex-col gap-4"
  >
    <div
      class="flex items-center"
    >
      <div
        class="rounded-full flex justify-center items-center  w-11 h-11"
      >
        <img
          alt="user profile"
          class="bg-white object-cover rounded-full p-[0.1rem] w-[42px] h-[42px] p-[0.1rem]"
          data-nimg="1"
          decoding="async"
          height="30"
          referrerpolicy="no-referrer"
          src=""
          style="color: transparent;"
          width="30"
        />
      </div>
      <div
        class="flex flex-col ml-5"
      >
        <p
          class="font-bold"
        >
          test
        </p>
        <p>
          test
        </p>
      </div>
    </div>
    <div>
      <p
        class="text-neutral-500"
      >
        About ∙ Help ∙ Press ∙ API ∙ Jobs ∙ 
        <br />
        Privacy ∙ Terms ∙ Location ∙ 
        <br />
        Language
      </p>
    </div>
    <div>
      <p
        class="text-neutral-500 font-bold"
      >
        @Copyright INSTANTGRAM 
        <br />
         from METAL
      </p>
    </div>
  </section>
</div>
`;
